File: src/Butil/Bit.Butil/Scripts/ink.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    const _trails: { [id: string]: { element: Element, handler: (event: PointerEvent) => void, style: any } } = {};
    // Requesting a presenter is asynchronous, so a stop() can land while one is still being
    // requested. The token says which start is still the current one for an id - an identity rather
    // than a counter, so that stop() can remove the entry outright instead of keeping a number
    // alive for every id the page has ever drawn with.
    const _starts: { [id: string]: object } = {};

    butil.ink = {
        isSupported() { return typeof (window.navigator as any).ink?.requestPresenter === 'function'; },

        // The presenter has to be fed the browser's own PointerEvent object, which is exactly what
        // cannot cross the interop boundary - by the time .NET saw one it would be a copy, and an
        // untrusted one at that. So the listener lives here and .NET only turns it on and off.
        async start(id: string, element: Element, color: string, diameter: number) {
            const ink = (window.navigator as any).ink;
            if (typeof ink?.requestPresenter !== 'function' || !element) return false;

            butil.ink.stop(id);

            const token = _starts[id] = {};

            let presenter: any;
            try {
                presenter = await ink.requestPresenter({ presentationArea: element });
            } catch {
                // Delegated ink is off, or the element is not a valid presentation area. Nothing is
                // in flight for this id any more unless a newer start has already claimed it.
                if (_starts[id] === token) delete _starts[id];
                return false;
            }

            // A stop() - or another start() - ran while the presenter was on its way: this one is
            // stale, and registering its listener would draw a trail nobody asked for any more.
            if (_starts[id] !== token) return false;

            const style = { color: color || 'black', diameter: diameter > 0 ? diameter : 3 };
            const handler = (event: PointerEvent) => {
                try { presenter.updateInkTrailStartPoint(event, style); } catch { /* the presenter went away */ }
            };

            _trails[id] = { element, handler, style };
            element.addEventListener('pointermove', handler as EventListener);
            return true;
        },

        // The style object is the same one the handler passes on every move, so mutating it changes
        // the trail from the next point onwards without re-requesting a presenter.
        setStyle(id: string, color: string, diameter: number) {
            const trail = _trails[id];
            if (!trail) return false;
            trail.style.color = color || trail.style.color;
            trail.style.diameter = diameter > 0 ? diameter : trail.style.diameter;
            return true;
        },

        stop(id: string) {
            // Dropped even with no trail to remove: a start still awaiting its presenter compares
            // its own token against this entry, and finding it gone is how it sees it was stopped.
            delete _starts[id];

            const trail = _trails[id];
            if (!trail) return;
            delete _trails[id];
            trail.element.removeEventListener('pointermove', trail.handler as EventListener);
        },
        disposeAll() {
            for (const id of Object.keys(_starts)) butil.ink.stop(id);
            for (const id of Object.keys(_trails)) butil.ink.stop(id);
        }
    };
}(BitButil));

File: src/Butil/Bit.Butil/Publics/Ink.cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/Ink_API">Ink API</see>
/// (delegated ink trails): while the user draws, the compositor paints ahead of your rendering so the
/// line keeps up with the stylus.
/// </summary>
/// <remarks>
/// A drawing app renders a stroke when its input event reaches the app, which is at least a frame
/// behind the pointer - visible as a gap between the stylus tip and the ink. Delegated ink hands the
/// last known point to the compositor, which draws the missing segment itself, outside the app's
/// frame loop. The app still draws the real stroke; this only covers the gap until it does.
/// <br/>
/// The presenter has to be fed the browser's own <c>PointerEvent</c>, which is exactly what cannot
/// cross the interop boundary - by the time .NET saw one it would be a copy, and an untrusted one.
/// So the listener lives in JavaScript and C# only turns it on and off. That is also why this is a
/// trail you start on an element rather than a method you call per point.
/// <br/>
/// Chromium-only, and an enhancement by nature: when it is unavailable nothing is lost but the few
/// milliseconds of latency it was hiding.
/// </remarks>
[ButilService(typeof(Ink))]
public class Ink(IJSRuntime js) : IAsyncDisposable
{
    /// <summary>True when the runtime exposes <c>navigator.ink.requestPresenter</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.ink.isSupported");

    /// <summary>
    /// Starts painting a delegated ink trail over an element as the pointer moves across it.
    /// </summary>
    /// <param name="presentationArea">
    /// The element the trail is drawn within - usually the canvas you are drawing on. The trail is
    /// clipped to it.
    /// </param>
    /// <param name="color">The trail's colour, as any CSS colour string.</param>
    /// <param name="diameter">The trail's width in pixels. Match it to your own stroke width, or the seam shows.</param>
    /// <returns>A handle - dispose it to stop - or null when the API is unavailable or the element can't present ink.</returns>
    /// <remarks>
    /// Keep drawing your own strokes as before. This does not replace your rendering; it only fills
    /// the gap between the pointer and the last frame you drew.
    /// </remarks>
    public async ValueTask<InkTrailHandle?> StartTrail(ElementReference presentationArea, string color = "black", double diameter = 3)
    {
        var id = Guid.NewGuid();
        var started = await js.Invoke<bool>("BitButil.ink.start", id, presentationArea, color, diameter);
        return started ? new InkTrailHandle(js, id) : null;
    }

    /// <summary>
    /// On scope/circuit teardown, stops any trail whose <see cref="InkTrailHandle"/> was never
    /// disposed, so an abandoned pointer listener can't outlive the component that started it.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            await js.InvokeVoid("BitButil.ink.disposeAll");
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed

        GC.SuppressFinalize(this);
    }
}
